mlir/lib/Quantum/Transforms/ChainedSelfInversePatterns.cpp

```diff
@@ -40,35 +44,109 @@ struct ChainedHadamardOpRewritePattern : public mlir::OpRewritePattern<CustomOp>
     {
         LLVM_DEBUG(dbgs() << "Simplifying the following operation:\n" << op << "\n");
 
+        VerifyParentGateAnalysis<CustomOp> vpga(op);
+        if (!vpga.getVerifierResult()) {
+            return failure();
+        }
+
         StringRef OpGateName = op.getGateName();
-        if (!HermitianOps.contains(OpGateName))
+        if (!HermitianOps.contains(OpGateName)) {
```

Comment:
To move before the verification

Reply:
Makes sense, this is much cheaper than verification